Subject:       Support Clang/LLVM for LTO and CFI for the kernel only
Patch Author:  Orson Teodoro <orsonteodoro@hotmail.com>
Patch Status:  Testing / Experimental
Date:          Thu Sep  9 12:03:03 PM PDT 2021 (Unix time: 1631214183)

v2 - Added utils support (experimental)
     Changed --clang to --clang-kernel
     Changed --llvm to --llvm-kernel
     Added --llvm-utils
     Added --clang-utils
     Dropped --pgo
     See genkernel-4.0.x/json-c/0.13.1/json-c-0.13.1-clang.patch folder for json-c patch
     Force busybox to be built with gcc
     Force lvm2 to be built with gcc
     Use llvm-ranlib
v1.1 Utilize more of the llvm toolchain
v1   Added llvm support

diff -urp genkernel-4.0.10.orig/arch/ppc64le/config.sh genkernel-4.0.10/arch/ppc64le/config.sh
--- genkernel-4.0.10.orig/arch/ppc64le/config.sh	2020-07-20 09:05:15.000000000 -0700
+++ genkernel-4.0.10/arch/ppc64le/config.sh	2021-09-10 01:49:49.439423278 -0700
@@ -18,9 +18,34 @@ DEFAULT_KERNEL_MAKE=make
 DEFAULT_UTILS_MAKE=make
 
 DEFAULT_KERNEL_CC=gcc
+DEFAULT_KERNEL_AR=ar
 DEFAULT_KERNEL_AS=as
-DEFAULT_KERNEL_LD=ld
+DEFAULT_KERNEL_LD=ld.bfd
+DEFAULT_KERNEL_NM=nm
+DEFAULT_KERNEL_OBJCOPY=objcopy
+DEFAULT_KERNEL_OBJDUMP=objdump
+DEFAULT_KERNEL_RANLIB=ranlib
+DEFAULT_KERNEL_READELF=readelf
+DEFAULT_KERNEL_STRIP=strip
+
+DEFAULT_KERNEL_LLVM_CC=clang
+DEFAULT_KERNEL_LLVM_AR=llvm-ar
+DEFAULT_KERNEL_LLVM_AS=llvm-as
+DEFAULT_KERNEL_LLVM_CLANG_FLAGS="-integrated-as"
+DEFAULT_KERNEL_LLVM_LD=ld.lld
+DEFAULT_KERNEL_LLVM_NM=llvm-nm
+DEFAULT_KERNEL_LLVM_OBJCOPY=llvm-objcopy
+DEFAULT_KERNEL_LLVM_OBJDUMP=llvm-objdump
+DEFAULT_KERNEL_LLVM_RANLIB=llvm-ranlib
+DEFAULT_KERNEL_LLVM_READELF=llvm-readelf
+DEFAULT_KERNEL_LLVM_STRIP=llvm-strip
 
 DEFAULT_UTILS_CC=gcc
+DEFAULT_UTILS_CXX=g++
 DEFAULT_UTILS_AS=as
 DEFAULT_UTILS_LD=ld
+
+DEFAULT_UTILS_LLVM_CC=clang
+DEFAULT_UTILS_LLVM_CXX=clang++
+DEFAULT_UTILS_LLVM_AS=llvm-as
+DEFAULT_UTILS_LLVM_LD=ld.lld
diff -urp genkernel-4.0.10.orig/defaults/config.sh genkernel-4.0.10/defaults/config.sh
--- genkernel-4.0.10.orig/defaults/config.sh	2020-07-20 09:05:15.000000000 -0700
+++ genkernel-4.0.10/defaults/config.sh	2021-09-10 01:49:49.440423317 -0700
@@ -41,8 +41,27 @@ DEFAULT_KERNEL_MAKE=make
 DEFAULT_UTILS_MAKE=make
 
 DEFAULT_KERNEL_CC=gcc
+DEFAULT_KERNEL_AR=ar
 DEFAULT_KERNEL_AS=as
-DEFAULT_KERNEL_LD=ld
+DEFAULT_KERNEL_LD=ld.bfd
+DEFAULT_KERNEL_NM=nm
+DEFAULT_KERNEL_OBJCOPY=objcopy
+DEFAULT_KERNEL_OBJDUMP=objdump
+DEFAULT_KERNEL_RANLIB=ranlib
+DEFAULT_KERNEL_READELF=readelf
+DEFAULT_KERNEL_STRIP=strip
+
+DEFAULT_KERNEL_LLVM_CC=clang
+DEFAULT_KERNEL_LLVM_AR=llvm-ar
+DEFAULT_KERNEL_LLVM_AS=llvm-as
+DEFAULT_KERNEL_LLVM_CLANG_FLAGS="-integrated-as"
+DEFAULT_KERNEL_LLVM_LD=ld.lld
+DEFAULT_KERNEL_LLVM_NM=llvm-nm
+DEFAULT_KERNEL_LLVM_OBJCOPY=llvm-objcopy
+DEFAULT_KERNEL_LLVM_OBJDUMP=llvm-objdump
+DEFAULT_KERNEL_LLVM_RANLIB=llvm-ranlib
+DEFAULT_KERNEL_LLVM_READELF=llvm-readelf
+DEFAULT_KERNEL_LLVM_STRIP=llvm-strip
 
 DEFAULT_UTILS_CFLAGS="-Os -pipe -fomit-frame-pointer"
 DEFAULT_UTILS_CC=gcc
@@ -50,5 +69,10 @@ DEFAULT_UTILS_CXX=g++
 DEFAULT_UTILS_AS=as
 DEFAULT_UTILS_LD=ld
 
+DEFAULT_UTILS_LLVM_CC=clang
+DEFAULT_UTILS_LLVM_CXX=clang++
+DEFAULT_UTILS_LLVM_AS=llvm-as
+DEFAULT_UTILS_LLVM_LD=ld.lld
+
 PORTAGE_CHOST="$(portageq envvar CHOST)"
 DEFAULT_CHOST="${PORTAGE_CHOST:-$(${DEFAULT_UTILS_CC} -dumpmachine 2>/dev/null)}"
diff -urp genkernel-4.0.10.orig/gen_cmdline.sh genkernel-4.0.10/gen_cmdline.sh
--- genkernel-4.0.10.orig/gen_cmdline.sh	2020-07-20 09:05:15.000000000 -0700
+++ genkernel-4.0.10/gen_cmdline.sh	2021-09-10 01:49:49.440423317 -0700
@@ -85,6 +85,9 @@ longusage() {
   echo "	--module-prefix=<dir>	Prefix to kernel module destination, modules"
   echo "				will be installed in <prefix>/lib/modules"
   echo "  Low-Level Compile settings"
+  echo "	--cfi			Use LLVM to build a CFI protected kernel"
+  echo "	--clang-kernel		Use LLVM to build the kernel"
+  echo "	--clang-utils		Use LLVM to build utils (Experimental / WIP / Debugging)"
   echo "	--cross-compile=<target-triplet>"
   echo "				Target triple (i.e. aarch64-linux-gnu) to build for"
   echo "	--kernel-as=<assembler>	Assembler to use for kernel"
@@ -109,6 +112,9 @@ longusage() {
   echo "	--nice			Run the kernel make at the default nice level (10)"
   echo "	--nice=<0-19>		Run the kernel make at the selected nice level"
   echo "	--no-nice		Don't be nice while running the kernel make"
+  echo "	--llvm-kernel		Use LLVM instead of GCC to compile the kernel"
+  echo "	--llvm-utils		Use LLVM instead of GCC to compile utils (Experimental / WIP / Debugging)"
+  echo "	--lto			Use LLVM to build a LTO optimized kernel"
   echo "  Initialization"
   echo "	--splash=<theme>	Enable framebuffer splash using <theme>"
   echo "	--splash-res=<res>	Select splash theme resolutions to install"
@@ -815,6 +821,14 @@ parse_cmdline() {
 		--config=*)
 			print_info 3 "CMD_GK_CONFIG: "${*#*=}""
 			;;
+		--cfi|--clang|--clang-kernel|--llvm|--llvm-kernel|--lto)
+			CMD_LLVM_KERNEL=1
+			print_info 3 "CMD_LLVM_KERNEL: ${CMD_LLVM_KERNEL}"
+			;;
+		--clang-utils|--llvm-utils)
+			CMD_LLVM_UTILS=1
+			print_info 3 "CMD_LLVM_UTILS: ${CMD_LLVM_UTILS}"
+			;;
 		--nice)
 			CMD_NICE=10
 			print_info 3 "CMD_NICE: ${CMD_NICE}"
diff -urp genkernel-4.0.10.orig/gen_compile.sh genkernel-4.0.10/gen_compile.sh
--- genkernel-4.0.10.orig/gen_compile.sh	2020-07-20 09:05:15.000000000 -0700
+++ genkernel-4.0.10/gen_compile.sh	2021-09-10 01:49:49.440423317 -0700
@@ -107,11 +107,17 @@ compile_generic() {
 			compile_cmd+=( "ARCH='${KERNEL_ARCH}'" )
 
 			local tc_var
-			for tc_var in AS AR CC LD NM OBJCOPY OBJDUMP READELF STRIP
+			for tc_var in AS AR CC LD NM OBJCOPY OBJDUMP RANLIB READELF STRIP
 			do
 				compile_cmd+=( "${tc_var}='$(TC_PROG_TYPE=KERNEL tc-get${tc_var})'" )
 			done
 
+			if [[ -n "${CMD_LLVM_KERNEL}" && "${CMD_LLVM_KERNEL}" == "1" ]] ; then
+				local max_llvm=$(ls "${BROOT}/usr/lib/llvm" | sort -rV | head -n 1)
+				export PATH="${BROOT}/usr/lib/llvm/${max_llvm}/bin:${PATH}"
+				compile_cmd+=( "CLANG_FLAGS='${CMD_KERNEL_CLANG_FLAGS}'" )
+			fi
+
 			compile_cmd+=( "HOSTAR='$(tc-getBUILD_AR)'" )
 			compile_cmd+=( "HOSTCC='$(tc-getBUILD_CC)'" )
 			compile_cmd+=( "HOSTCXX='$(tc-getBUILD_CXX)'" )
diff -urp genkernel-4.0.10.orig/gen_determineargs.sh genkernel-4.0.10/gen_determineargs.sh
--- genkernel-4.0.10.orig/gen_determineargs.sh	2020-07-20 09:05:15.000000000 -0700
+++ genkernel-4.0.10/gen_determineargs.sh	2021-09-10 01:49:49.441423356 -0700
@@ -258,6 +258,26 @@ determine_output_filenames() {
 	done
 }
 
+kernel_default() {
+	local gcc_default="${1}"
+	local llvm_default="${2}"
+	if [[ -n "${CMD_LLVM_LLVM}" && "${CMD_LLVM_KERNEL}" == "1" ]] ; then
+		echo "${llvm_default}"
+	else
+		echo "${gcc_default}"
+	fi
+}
+
+utils_default() {
+	local gcc_default="${1}"
+	local llvm_default="${2}"
+	if [[ -n "${CMD_LLVM_UTILS}" && "${CMD_LLVM_UTILS}" == "1" ]] ; then
+		echo "${llvm_default}"
+	else
+		echo "${gcc_default}"
+	fi
+}
+
 determine_real_args() {
 	# Unset known variables which will interfere with _tc-getPROG().
 	local tc_var tc_varname_build tc_vars=$(get_tc_vars)
@@ -306,13 +326,21 @@ determine_real_args() {
 	set_config_with_override STRING KERNEL_MAKE                           CMD_KERNEL_MAKE                           "${DEFAULT_KERNEL_MAKE}"
 	set_config_with_override STRING UTILS_CFLAGS                          CMD_UTILS_CFLAGS                          "${DEFAULT_UTILS_CFLAGS}"
 	set_config_with_override STRING UTILS_MAKE                            CMD_UTILS_MAKE                            "${DEFAULT_UTILS_MAKE}"
-	set_config_with_override STRING KERNEL_CC                             CMD_KERNEL_CC                             "${DEFAULT_KERNEL_CC}"
-	set_config_with_override STRING KERNEL_LD                             CMD_KERNEL_LD                             "${DEFAULT_KERNEL_LD}"
-	set_config_with_override STRING KERNEL_AS                             CMD_KERNEL_AS                             "${DEFAULT_KERNEL_AS}"
-	set_config_with_override STRING UTILS_CC                              CMD_UTILS_CC                              "${DEFAULT_UTILS_CC}"
-	set_config_with_override STRING UTILS_CXX                             CMD_UTILS_CXX                             "${DEFAULT_UTILS_CXX}"
-	set_config_with_override STRING UTILS_LD                              CMD_UTILS_LD                              "${DEFAULT_UTILS_LD}"
-	set_config_with_override STRING UTILS_AS                              CMD_UTILS_AS                              "${DEFAULT_UTILS_AS}"
+	set_config_with_override STRING KERNEL_CC                             CMD_KERNEL_CC                             $(kernel_default "${DEFAULT_KERNEL_CC}" "${DEFAULT_KERNEL_LLVM_CC}")
+	set_config_with_override STRING KERNEL_CLANG_FLAGS                    CMD_KERNEL_CLANG_FLAGS                    "${DEFAULT_KERNEL_LLVM_CLANG_FLAGS}"
+	set_config_with_override STRING KERNEL_LD                             CMD_KERNEL_LD                             $(kernel_default "${DEFAULT_KERNEL_LD}" "${DEFAULT_KERNEL_LLVM_LD}")
+	set_config_with_override STRING KERNEL_AR                             CMD_KERNEL_AR                             $(kernel_default "${DEFAULT_KERNEL_AR}" "${DEFAULT_KERNEL_LLVM_AR}")
+	set_config_with_override STRING KERNEL_AS                             CMD_KERNEL_AS                             $(kernel_default "${DEFAULT_KERNEL_AS}" "${DEFAULT_KERNEL_LLVM_AS}")
+	set_config_with_override STRING KERNEL_NM                             CMD_KERNEL_NM                             $(kernel_default "${DEFAULT_KERNEL_NM}" "${DEFAULT_KERNEL_LLVM_NM}")
+	set_config_with_override STRING KERNEL_OBJCOPY                        CMD_KERNEL_OBJCOPY                        $(kernel_default "${DEFAULT_KERNEL_OBJCOPY}" "${DEFAULT_KERNEL_LLVM_OBJCOPY}")
+	set_config_with_override STRING KERNEL_OBJDUMP                        CMD_KERNEL_OBJDUMP                        $(kernel_default "${DEFAULT_KERNEL_OBJDUMP}" "${DEFAULT_KERNEL_LLVM_OBJDUMP}")
+	set_config_with_override STRING KERNEL_RANLIB                         CMD_KERNEL_RANLIB                         $(kernel_default "${DEFAULT_KERNEL_RANLIB}" "${DEFAULT_KERNEL_LLVM_RANLIB}")
+	set_config_with_override STRING KERNEL_READELF                        CMD_KERNEL_READELF                        $(kernel_default "${DEFAULT_KERNEL_READELF}" "${DEFAULT_KERNEL_LLVM_READELF}")
+	set_config_with_override STRING KERNEL_STRIP                          CMD_KERNEL_STRIP                          $(kernel_default "${DEFAULT_KERNEL_STRIP}" "${DEFAULT_KERNEL_LLVM_STRIP}")
+	set_config_with_override STRING UTILS_CC                              CMD_UTILS_CC                              $(utils_default "${DEFAULT_UTILS_CC}" "${DEFAULT_UTILS_LLVM_CC}")
+	set_config_with_override STRING UTILS_CXX                             CMD_UTILS_CXX                             $(utils_default "${DEFAULT_UTILS_CXX}" "${DEFAULT_UTILS_LLVM_CXX}")
+	set_config_with_override STRING UTILS_LD                              CMD_UTILS_LD                              $(utils_default "${DEFAULT_UTILS_LD}" "${DEFAULT_UTILS_LLVM_LD}")
+	set_config_with_override STRING UTILS_AS                              CMD_UTILS_AS                              $(utils_default "${DEFAULT_UTILS_AS}" "${DEFAULT_UTILS_LLVM_AS}")
 
 	set_config_with_override STRING CROSS_COMPILE                         CMD_CROSS_COMPILE
 	set_config_with_override STRING BOOTDIR                               CMD_BOOTDIR                               "/boot"
diff -urp genkernel-4.0.10.orig/genkernel genkernel-4.0.10/genkernel
--- genkernel-4.0.10.orig/genkernel	2020-07-20 09:05:15.000000000 -0700
+++ genkernel-4.0.10/genkernel	2021-09-10 01:50:58.701117662 -0700
@@ -364,6 +364,11 @@ if isTrue "${BUILD_RAMDISK}"
 then
 	print_info 1 '' 1 0
 
+	if [[ -n "${CMD_LLVM_UTILS}" && "${CMD_LLVM_UTILS}" == "1" ]] ; then
+		local max_llvm=$(ls "${BROOT}/usr/lib/llvm" | sort -rV | head -n 1)
+		export PATH="${BROOT}/usr/lib/llvm/${max_llvm}/bin:${PATH}"
+	fi
+
 	# Compile initramfs
 	create_initramfs
 else
diff -urp genkernel-4.0.10.orig/gkbuilds/busybox.gkbuild genkernel-4.0.10/gkbuilds/busybox.gkbuild
--- genkernel-4.0.10.orig/gkbuilds/busybox.gkbuild	2020-07-20 09:05:15.000000000 -0700
+++ genkernel-4.0.10/gkbuilds/busybox.gkbuild	2021-09-10 01:49:49.441423356 -0700
@@ -1,6 +1,26 @@
 # Copyright 1999-2020 Gentoo Authors
 # Distributed under the terms of the GNU General Public License v2
 
+use_gcc_toolchain() {
+	# This package does not work with clang/llvm toolchain.
+	# Reason:  Builds fine but segfaults or panics.
+	export CC=gcc
+	export CXX=g++
+	export AS=as
+	export AR=ar
+	export LD=ld.bfd
+	export NM=nm
+	export OBJCOPY=objcopy
+	export OBJDUMP=objdump
+	export RANLIB=ranlib
+	export READELF=readelf
+	export STRIP=strip
+	export HOSTAR=${CHOST}-ar
+	export HOSTCC=${CHOST}-gcc
+	export HOSTCXX=${CHOST}-g++
+	export HOSTLD=${CHOST}-ld.bfd
+}
+
 src_prepare() {
 	default
 
@@ -33,6 +53,7 @@ src_configure() {
 }
 
 src_install() {
+	use_gcc_toolchain
 	#gkmake CONFIG_PREFIX="${D}" install
 	mkdir "${D}"/bin || die "Failed to create '${D}/bin'!"
 
diff -urp genkernel-4.0.10.orig/gkbuilds/lvm.gkbuild genkernel-4.0.10/gkbuilds/lvm.gkbuild
--- genkernel-4.0.10.orig/gkbuilds/lvm.gkbuild	2020-07-20 09:05:15.000000000 -0700
+++ genkernel-4.0.10/gkbuilds/lvm.gkbuild	2021-09-10 01:49:49.441423356 -0700
@@ -1,7 +1,28 @@
 # Copyright 1999-2019 Gentoo Authors
 # Distributed under the terms of the GNU General Public License v2
 
+use_gcc_toolchain() {
+	# This package does not work with clang/llvm toolchain.
+	# Reason:  Build time failure
+	export CC=gcc
+	export CXX=g++
+	export AS=as
+	export AR=ar
+	export LD=ld.bfd
+	export NM=nm
+	export OBJCOPY=objcopy
+	export OBJDUMP=objdump
+	export RANLIB=ranlib
+	export READELF=readelf
+	export STRIP=strip
+	export HOSTAR=${CHOST}-ar
+	export HOSTCC=${CHOST}-gcc
+	export HOSTCXX=${CHOST}-g++
+	export HOSTLD=${CHOST}-ld.bfd
+}
+
 src_configure() {
+	use_gcc_toolchain
 
 	append-ldflags -Wl,-rpath-link,${BROOT}/usr/lib
 
